File: ios-app/SafeTMobile/Audio/VoiceLinkTelemetryReporter.swift
import Foundation
import os

/// Per-window inbound voice-link counters reported every ~30 s.
///
/// Mirrors the Android / web reporters so all three clients post the same
/// payload schema to `POST /v1/telemetry/voice-link`. Counters only — no audio,
/// no transcript, no PCM. The admin dashboard reads aggregates back so
/// dispatch can answer "is this unit having voice quality problems?" with data
/// instead of trusting an end-user report.
///
/// Lifecycle:
///   - `start(baseURL:tokenProvider:)` arms a Timer that rolls the in-progress
///     window into a queued snapshot every 30 s and POSTs the head.
///   - `stop()` invalidates the timer. Counters carry across `stop()` /
///     `start()` into the next window — no data lost across an app pause.
///
/// Buffering:
///   - Up to `maxBufferedWindows` (~2 minutes) of unsent windows in memory.
///   - On a failed POST the head is left at the front of the queue; the next
///     30-second tick retries it. When the cap is exceeded the OLDEST queued
///     window is dropped, not the newest — operators care about recent data
///     first.
///   - Idle windows (zero counters) are sent too so the dashboard can tell
///     "this unit is alive but quiet" apart from "this unit fell off the air".
///
/// Thread safety: counters are mutated from `VoiceTransport.dispatchInboundVoice`
/// (the main actor) and `InboundJitterBuffer.playoutLoop` (a dedicated thread).
/// All accesses go through `lock`; the snapshot path closes the in-flight
/// window under the same lock.
final class VoiceLinkTelemetryReporter {

    static let shared = VoiceLinkTelemetryReporter()

    private static let telemetryIntervalSeconds: TimeInterval = 30
    private static let maxBufferedWindows = 4
    private static let clientType = "ios"

    private let lock = NSLock()
    // Use the explicit class name (not `Self`) here: Swift rejects covariant
    // `Self` in a stored-property default initializer because the initializer
    // expression is evaluated before the type is fully formed.
    private var window = WindowCounters(openedAtMs: VoiceLinkTelemetryReporter.nowMs())
    private var queued: [QueuedWindow] = []
    private var timer: Timer?
    private var inFlight = false

    private var baseURL: URL?
    private var tokenProvider: () -> String = { "" }
    private var radioKeyProvider: () -> String = { "" }
    private var unitId: String?
    private var channel: String?

    private let logger = Logger(subsystem: "com.safetptt.mobile", category: "voiceLinkTelemetry")
    private let session: URLSession = {
        let cfg = URLSessionConfiguration.ephemeral
        cfg.timeoutIntervalForRequest = 15
        cfg.timeoutIntervalForResource = 15
        cfg.waitsForConnectivity = false
        return URLSession(configuration: cfg)
    }()

    /// Sets the POST destination + auth providers. Safe to call repeatedly;
    /// only the latest values are used on the next POST.
    func configure(baseURL: URL, tokenProvider: @escaping () -> String, radioKeyProvider: @escaping () -> String) {
        lock.lock()
        self.baseURL = baseURL
        self.tokenProvider = tokenProvider
        self.radioKeyProvider = radioKeyProvider
        lock.unlock()
    }

    /// Sets which (unit, channel) future reports are billed to. Calling with a
    /// new identity closes the in-progress window under the previous identity
    /// so counters accumulated on the prior channel are not credited to the
    /// new one.
    func setIdentity(unitId: String?, channel: String?) {
        lock.lock()
        let identityChanged = self.unitId != unitId || self.channel != channel
        if identityChanged, self.unitId != nil {
            closeAndQueueWindowLocked()
        }
        self.unitId = unitId
        self.channel = channel
        lock.unlock()
    }

    /// Arms the 30-second POST loop. Idempotent — a second `start` while
    /// already running is a no-op.
    func start() {
        lock.lock()
        if timer != nil {
            lock.unlock()
            return
        }
        // Timer must be created on a run loop. Use the main run loop because
        // VoiceTransport is `@MainActor` and most callers will already be on
        // the main thread when they call start().
        let t = Timer(timeInterval: Self.telemetryIntervalSeconds, repeats: true) { [weak self] _ in
            self?.tick()
        }
        timer = t
        lock.unlock()
        DispatchQueue.main.async {
            RunLoop.main.add(t, forMode: .common)
        }
    }

    func stop() {
        lock.lock()
        let t = timer
        timer = nil
        lock.unlock()
        // Timer.invalidate must run on the run loop the timer was added to;
        // we hop to main to match the add path above, otherwise the call
        // becomes a no-op and the timer fires once more before the loop
        // notices it's invalid.
        if let t {
            DispatchQueue.main.async {
                t.invalidate()
            }
        }
    }

    // MARK: - counter recording

    func recordFrameReceived(codec: String, bytes: Int) {
        lock.lock()
        window.framesReceived += 1
        window.bytesReceived += max(0, bytes)
        var entry = window.codecBreakdown[codec] ?? CodecCounters()
        entry.framesReceived += 1
        window.codecBreakdown[codec] = entry
        lock.unlock()
    }

    /// Uplink accounting — every app-level byte this handset puts on the voice
    /// socket (vocoded frames, clear PCM, recorder sideband) so the admin
    /// data-usage column reflects both directions.
    func recordBytesSent(_ bytes: Int) {
        lock.lock()
        window.bytesSent += max(0, bytes)
        lock.unlock()
    }

    func recordFrameDecoded(codec: String) {
        lock.lock()
        window.framesDecoded += 1
        var entry = window.codecBreakdown[codec] ?? CodecCounters()
        entry.framesDecoded += 1
        window.codecBreakdown[codec] = entry
        lock.unlock()
    }

    func recordDecodeFailure() {
        lock.lock()
        window.decodeFailures += 1
        lock.unlock()
    }

    func recordPlcSynthesized() {
        lock.lock()
        window.plcFramesSynthesized += 1
        lock.unlock()
    }

    func recordBufferUnderrun() {
        lock.lock()
        window.bufferUnderruns += 1
        lock.unlock()
    }

    func recordBufferDepth(_ frames: Int) {
        lock.lock()
        if frames > window.maxBufferDepthFrames {
            window.maxBufferDepthFrames = frames
        }
        lock.unlock()
    }

    func recordTalkSpurtStart() {
        lock.lock()
        window.talkSpurtsStarted += 1
        lock.unlock()
    }

    func recordTalkSpurtEnd() {
        lock.lock()
        window.talkSpurtsEnded += 1
        lock.unlock()
    }

    // MARK: - private

    private func tick() {
        lock.lock()
        guard unitId != nil else {
            // No identity yet — rotate the window so accumulated counters
            // don't backfill onto the first identity that arrives.
            window = WindowCounters(openedAtMs: Self.nowMs())
            lock.unlock()
            return
        }
        closeAndQueueWindowLocked()
        lock.unlock()
        flush()
    }

    private func closeAndQueueWindowLocked() {
        guard let u = unitId else { return }
        let closedAt = Self.nowMs()
        queued.append(QueuedWindow(unitId: u, channel: channel, counters: window, closedAtMs: closedAt))
        window = WindowCounters(openedAtMs: closedAt)
        // Drop the OLDEST queued window when the cap is exceeded — operators
        // care about recent data first; an hour-old summary is less
        // actionable than the current one.
        while queued.count > Self.maxBufferedWindows {
            queued.removeFirst()
        }
    }

    private func flush() {
        lock.lock()
        if inFlight {
            lock.unlock()
            return
        }
        inFlight = true
        lock.unlock()
        defer {
            lock.lock()
            inFlight = false
            lock.unlock()
        }
        while true {
            lock.lock()
            guard let head = queued.first, let url = baseURL else {
                lock.unlock()
                return
            }
            let token = tokenProvider().trimmingCharacters(in: .whitespacesAndNewlines)
            let radioKey = radioKeyProvider().trimmingCharacters(in: .whitespacesAndNewlines)
            lock.unlock()
            let ok = postSync(url: url, window: head, token: token, radioKey: radioKey)
            if !ok { return }
            lock.lock()
            if let first = queued.first, first.closedAtMs == head.closedAtMs && first.unitId == head.unitId {
                queued.removeFirst()
            }
            lock.unlock()
        }
    }

    private func postSync(url: URL, window: QueuedWindow, token: String, radioKey: String) -> Bool {
        var request = URLRequest(url: url.appendingPathComponent("/v1/telemetry/voice-link"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        } else if !radioKey.isEmpty {
            request.setValue(radioKey, forHTTPHeaderField: "X-Radio-Key")
        }
        guard let body = Self.buildReportBody(window) else { return true /* drop bad body */ }
        request.httpBody = body
        let sem = DispatchSemaphore(value: 0)
        var success = false
        var permanent = false
        let task = session.dataTask(with: request) { _, response, _ in
            defer { sem.signal() }
            guard let http = response as? HTTPURLResponse else { return }
            // 2xx + 202 (DB-less soft-accept) drain the head; 5xx / network
            // errors retain it for retry; non-429 4xx is permanently broken
            // (bad payload, wrong auth) so drain to avoid burning retries.
            if (200..<300).contains(http.statusCode) {
                success = true
            } else if http.statusCode == 429 {
                success = false
            } else if (400..<500).contains(http.statusCode) {
                permanent = true
            }
        }
        task.resume()
        // Bounded wait — Timer fires every 30 s, so a stuck request can stall
        // at most one tick before we give up and let the next one try again.
        _ = sem.wait(timeout: .now() + .seconds(20))
        return success || permanent
    }

    // MARK: - shape (visible for tests)

    struct CodecCounters: Equatable {
        var framesReceived: Int = 0
        var framesDecoded: Int = 0
    }

    struct WindowCounters {
        let openedAtMs: Int64
        var framesReceived: Int = 0
        var framesDecoded: Int = 0
        var decodeFailures: Int = 0
        var plcFramesSynthesized: Int = 0
        var bufferUnderruns: Int = 0
        var maxBufferDepthFrames: Int = 0
        var talkSpurtsStarted: Int = 0
        var talkSpurtsEnded: Int = 0
        var bytesReceived: Int = 0
        /// Uplink bytes — voice frames + recorder sideband sent on the socket.
        var bytesSent: Int = 0
        var codecBreakdown: [String: CodecCounters] = [:]
    }

    struct QueuedWindow {
        let unitId: String
        let channel: String?
        let counters: WindowCounters
        let closedAtMs: Int64
    }

    /// Builds the JSON wire body. Pure function — exposed for tests.
    static func buildReportBody(_ w: QueuedWindow) -> Data? {
        var obj: [String: Any] = [
            "unitId": w.unitId,
            "clientType": Self.clientType,
            "counters": [
                "framesReceived": w.counters.framesReceived,
                "framesDecoded": w.counters.framesDecoded,
                "decodeFailures": w.counters.decodeFailures,
                "plcFramesSynthesized": w.counters.plcFramesSynthesized,
                "bufferUnderruns": w.counters.bufferUnderruns,
                "maxBufferDepthFrames": w.counters.maxBufferDepthFrames,
                "talkSpurtsStarted": w.counters.talkSpurtsStarted,
                "talkSpurtsEnded": w.counters.talkSpurtsEnded,
                "bytesReceived": w.counters.bytesReceived,
            "bytesSent": w.counters.bytesSent,
                "wallMsObservation": max(0, Int(w.closedAtMs - w.counters.openedAtMs)),
            ] as [String: Any],
            "clientTs": Self.isoFormat(ms: w.closedAtMs),
        ]
        if let channel = w.channel {
            obj["channel"] = channel
        }
        var codecBreakdown: [String: [String: Int]] = [:]
        for (codec, c) in w.counters.codecBreakdown {
            codecBreakdown[codec] = [
                "framesReceived": c.framesReceived,
                "framesDecoded": c.framesDecoded,
            ]
        }
        obj["codecBreakdown"] = codecBreakdown
        return try? JSONSerialization.data(withJSONObject: obj, options: [])
    }

    private static func nowMs() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000.0)
    }

    private static let isoFormatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static func isoFormat(ms: Int64) -> String {
        Self.isoFormatter.string(from: Date(timeIntervalSince1970: TimeInterval(ms) / 1000.0))
    }

    // Visible for tests
    func snapshotForTest() -> (queuedCount: Int, framesReceived: Int) {
        lock.lock()
        defer { lock.unlock() }
        return (queued.count, window.framesReceived)
    }

    func resetForTest() {
        lock.lock()
        queued.removeAll()
        window = WindowCounters(openedAtMs: Self.nowMs())
        unitId = nil
        channel = nil
        lock.unlock()
    }
}

File: android-app/app/src/main/java/com/securityradio/ptt/device/VoiceLinkTelemetryReporter.kt
package com.securityradio.ptt.device

import com.securityradio.ptt.BuildConfig
import com.securityradio.ptt.data.remote.normalizeApiBaseUrl
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.Locale
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * Per-window inbound voice-link counters reported every ~30 s.
 *
 * Mirrors the iOS / web reporters so all three clients post the same payload
 * schema to `POST /v1/telemetry/voice-link`. Counters only — no audio, no
 * transcript, no PCM. The dashboard reads aggregates back so dispatch can
 * answer "is this unit having voice quality problems?" with data instead of
 * trusting an end-user report.
 *
 * Lifecycle:
 *   - [start] arms a background coroutine that rolls the in-progress window
 *     into a queued snapshot every 30 s and POSTs the head.
 *   - [stop] cancels the loop. Counters carry across `stop()` / `start()`
 *     into the next window — no data lost across an app pause.
 *
 * Buffering:
 *   - Up to [MAX_BUFFERED_WINDOWS] (~2 minutes) of unsent windows in memory.
 *   - On a failed POST the head is left at the front of the queue and the
 *     next 30-second tick retries it. When the cap is exceeded the OLDEST
 *     queued window is dropped, not the newest — operators care about
 *     recent data first.
 *   - Idle windows (zero counters) are sent too so the dashboard can tell
 *     "this unit is alive but quiet" apart from "this unit fell off the air".
 *
 * Thread safety: counters are mutated from the WebSocket dispatch thread
 * (`VoiceRelayTransport.dispatchInboundVoice`) and from the jitter buffer
 * playout thread (`InboundJitterBuffer.playoutLoop`). All accesses go
 * through the synchronized counter helpers below; the snapshot path
 * captures the in-flight values under the same lock.
 */
object VoiceLinkTelemetryReporter {

    private const val TELEMETRY_INTERVAL_MS = 30_000L
    private const val MAX_BUFFERED_WINDOWS = 4
    private const val CLIENT_TYPE = "android"

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val running = AtomicBoolean(false)
    private var loopJob: Job? = null

    private val lock = ReentrantLock()
    /** In-progress window. Rolled into [queued] on every tick. */
    private var window: WindowCounters = WindowCounters(System.currentTimeMillis())
    private val queued = ArrayDeque<QueuedWindow>()
    private val inFlight = AtomicBoolean(false)

    @Volatile
    private var baseUrl: String? = null
    @Volatile
    private var apiKeyProvider: (() -> String) = { "" }
    @Volatile
    private var authTokenProvider: (() -> String) = { "" }
    @Volatile
    private var unitId: String? = null
    @Volatile
    private var channel: String? = null

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .writeTimeout(15, TimeUnit.SECONDS)
        .build()

    /**
     * Configures the POST destination + auth. Safe to call repeatedly; only
     * the latest values are used on the next POST.
     */
    fun configure(
        httpBaseUrl: String,
        authTokenProvider: () -> String,
        apiKeyProvider: () -> String,
    ) {
        this.baseUrl = normalizeApiBaseUrl(httpBaseUrl).trimEnd('/')
        this.authTokenProvider = authTokenProvider
        this.apiKeyProvider = apiKeyProvider
    }

    /**
     * Sets which unit/channel future reports are billed to. Calling with a
     * new identity closes the current window under the previous identity so
     * counters that were accumulated while on the prior channel are not
     * credited to the new one.
     */
    fun setIdentity(unitId: String?, channel: String?) {
        lock.withLock {
            val identityChanged = this.unitId != unitId || this.channel != channel
            if (identityChanged && this.unitId != null) {
                closeAndQueueWindowLocked()
            }
            this.unitId = unitId
            this.channel = channel
        }
    }

    fun start() {
        if (!running.compareAndSet(false, true)) return
        loopJob = scope.launch {
            while (running.get()) {
                delay(TELEMETRY_INTERVAL_MS)
                tick()
            }
        }
    }

    fun stop() {
        if (!running.compareAndSet(true, false)) return
        loopJob?.cancel()
        loopJob = null
    }

    // --- counter recording (call sites in VoiceRelayTransport / Inbound) --

    fun recordFrameReceived(codec: String, bytes: Int) {
        lock.withLock {
            window.framesReceived += 1
            window.bytesReceived += bytes.coerceAtLeast(0)
            val entry = window.codecBreakdown.getOrPut(codec) { CodecCounters() }
            entry.framesReceived += 1
        }
    }

    /**
     * Uplink accounting — every app-level byte this handset puts on the voice
     * socket (vocoded frames, clear PCM, recorder sideband) so the admin
     * data-usage column reflects both directions.
     */
    fun recordBytesSent(bytes: Int) {
        lock.withLock {
            window.bytesSent += bytes.coerceAtLeast(0)
        }
    }

    fun recordFrameDecoded(codec: String) {
        lock.withLock {
            window.framesDecoded += 1
            val entry = window.codecBreakdown.getOrPut(codec) { CodecCounters() }
            entry.framesDecoded += 1
        }
    }

    fun recordDecodeFailure() {
        lock.withLock {
            window.decodeFailures += 1
        }
    }

    fun recordPlcSynthesized() {
        lock.withLock {
            window.plcFramesSynthesized += 1
        }
    }

    fun recordBufferUnderrun() {
        lock.withLock {
            window.bufferUnderruns += 1
        }
    }

    fun recordBufferDepth(frames: Int) {
        lock.withLock {
            if (frames > window.maxBufferDepthFrames) {
                window.maxBufferDepthFrames = frames
            }
        }
    }

    fun recordTalkSpurtStart() {
        lock.withLock {
            window.talkSpurtsStarted += 1
        }
    }

    fun recordTalkSpurtEnd() {
        lock.withLock {
            window.talkSpurtsEnded += 1
        }
    }

    // --- internal --------------------------------------------------------

    private fun tick() {
        val identity = lock.withLock {
            val u = unitId ?: return
            // Identity captured under the lock so the queue read below sees
            // a consistent (unit, channel) for the just-closed window.
            closeAndQueueWindowLocked()
            u
        }
        if (identity.isBlank()) return
        flush()
    }

    private fun closeAndQueueWindowLocked() {
        val closedAt = System.currentTimeMillis()
        val u = unitId ?: return
        queued.addLast(
            QueuedWindow(
                unitId = u,
                channel = channel,
                counters = window,
                closedAtMs = closedAt,
            ),
        )
        window = WindowCounters(closedAt)
        // Pre-cap so a long offline period doesn't unbounded-grow memory.
        while (queued.size > MAX_BUFFERED_WINDOWS) {
            queued.removeFirst()
        }
    }

    private fun flush() {
        if (!inFlight.compareAndSet(false, true)) return
        try {
            val url = baseUrl ?: return
            val endpoint = "$url/v1/telemetry/voice-link"
            while (true) {
                val head = lock.withLock { queued.firstOrNull() } ?: break
                val ok = postWindow(endpoint, head)
                if (!ok) break
                lock.withLock {
                    // Defend against a concurrent identity change that already
                    // shifted the queue while the POST was inflight.
                    if (queued.isNotEmpty() && queued.first() === head) {
                        queued.removeFirst()
                    }
                }
            }
        } finally {
            inFlight.set(false)
        }
    }

    private fun postWindow(endpoint: String, w: QueuedWindow): Boolean {
        val bodyJson = buildReportBody(w)
        val requestBuilder = Request.Builder()
            .url(endpoint)
            .post(bodyJson.toRequestBody("application/json".toMediaType()))
        val token = authTokenProvider().trim()
        val key = apiKeyProvider().trim()
        if (token.isNotEmpty()) {
            requestBuilder.header("Authorization", "Bearer $token")
        } else if (key.isNotEmpty()) {
            requestBuilder.header("X-Radio-Key", key)
        }
        return try {
            httpClient.newCall(requestBuilder.build()).execute().use { response ->
                // 2xx + 202 (DB-less soft-accept) both drain the queue; only
                // 5xx / network errors retain the head for retry. 4xx other
                // than 429 means the client is permanently misshapen — drop
                // the head so we don't burn retries on a bad payload.
                when {
                    response.isSuccessful -> true
                    response.code == 429 -> false
                    response.code in 400..499 -> true
                    else -> false
                }
            }
        } catch (_: Exception) {
            false
        }
    }

    /** Visible for tests / debug logging. Pure builder of the JSON wire body. */
    internal fun buildReportBody(w: QueuedWindow): String {
        val obj = JSONObject()
        obj.put("unitId", w.unitId)
        if (w.channel != null) obj.put("channel", w.channel)
        obj.put("clientType", CLIENT_TYPE)
        // App version rides along so safeT Control can show each unit's build and
        // flag out-of-date radios (drives the fleet OTA view).
        obj.put("appVersionName", BuildConfig.VERSION_NAME)
        obj.put("appVersionCode", BuildConfig.VERSION_CODE)
        val counters = JSONObject()
        counters.put("framesReceived", w.counters.framesReceived)
        counters.put("framesDecoded", w.counters.framesDecoded)
        counters.put("decodeFailures", w.counters.decodeFailures)
        counters.put("plcFramesSynthesized", w.counters.plcFramesSynthesized)
        counters.put("bufferUnderruns", w.counters.bufferUnderruns)
        counters.put("maxBufferDepthFrames", w.counters.maxBufferDepthFrames)
        counters.put("talkSpurtsStarted", w.counters.talkSpurtsStarted)
        counters.put("talkSpurtsEnded", w.counters.talkSpurtsEnded)
        counters.put("bytesReceived", w.counters.bytesReceived)
        counters.put("bytesSent", w.counters.bytesSent)
        counters.put("wallMsObservation", (w.closedAtMs - w.counters.windowOpenedAtMs).coerceAtLeast(0))
        obj.put("counters", counters)
        val codecBreakdown = JSONObject()
        for ((codec, c) in w.counters.codecBreakdown) {
            val entry = JSONObject()
            entry.put("framesReceived", c.framesReceived)
            entry.put("framesDecoded", c.framesDecoded)
            codecBreakdown.put(codec, entry)
        }
        obj.put("codecBreakdown", codecBreakdown)
        obj.put("clientTs", java.text.SimpleDateFormat(
            "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'",
            Locale.US,
        ).apply {
            timeZone = java.util.TimeZone.getTimeZone("UTC")
        }.format(java.util.Date(w.closedAtMs)))
        return obj.toString()
    }

    // Visible for tests
    internal data class CodecCounters(
        var framesReceived: Int = 0,
        var framesDecoded: Int = 0,
    )

    // Visible for tests
    internal class WindowCounters(val windowOpenedAtMs: Long) {
        var framesReceived: Int = 0
        var framesDecoded: Int = 0
        var decodeFailures: Int = 0
        var plcFramesSynthesized: Int = 0
        var bufferUnderruns: Int = 0
        var maxBufferDepthFrames: Int = 0
        var talkSpurtsStarted: Int = 0
        var talkSpurtsEnded: Int = 0
        var bytesReceived: Int = 0
        var bytesSent: Int = 0
        val codecBreakdown: MutableMap<String, CodecCounters> = LinkedHashMap()
    }

    // Visible for tests
    internal data class QueuedWindow(
        val unitId: String,
        val channel: String?,
        val counters: WindowCounters,
        val closedAtMs: Long,
    )

    // Visible for tests: snapshot for assertions.
    internal fun queuedSizeForTest(): Int = lock.withLock { queued.size }
    internal fun currentWindowFramesReceivedForTest(): Int = lock.withLock { window.framesReceived }
    internal fun resetForTest() {
        lock.withLock {
            queued.clear()
            window = WindowCounters(System.currentTimeMillis())
            unitId = null
            channel = null
        }
    }

}
